### PR TITLE
Added `migration_status` table to capture a snapshot of migrated tables

### LINKED ISSUE #905: [FEATURE]: Check table migration progress
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem statement

If a user triggers a migration process remotely (from the CLI) they cannot tell the status of the migration

### Proposed Solution

Enable remote checking of the migration status in % or number of tables/total tables.

Create new table - `ucx.table_migration_progress`:

| src_workspace_id | src_schema | src_table | dst_catalog | dst_schema | dst_table | updated
| --- | --- | --- | --- | --- | --- |
| 1234 | foo_db | foo_table | migrated_catalog | foo | table | 2024-01-02 |
| 5678 | bar_db | bar_table | `NULL` | `NULL` | `NULL` | `NULL` |

We need to refactor the logic into `MigrationProgress` class from the `TableMigrate` class:
https://github.com/databrickslabs/ucx/blob/main/src/databricks/labs/ucx/hive_metastore/table_migrate.py#L112-L151
- it is supposed to be called every time we migrate a batch of external (or other tables)
- it is supposed to be called after we revert or skip the migration
- ...
- we'll use this table as the input for a new dashboard (out of scope of this issue)
- we'll use this table as the input for code migration flow

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by FastLee**
### Linked issues

Resolves #905

### Functionality 

- [ ] added relevant user documentation
- [ ] added new CLI command
- [ ] modified existing command: `databricks labs ucx ...`
- [ ] added a new workflow
- [ ] modified existing workflow: `...`
- [x] added a new table `table_migration_status`
- [ ] modified existing table: `...`

### Tests

- [ ] manually tested
- [ ] added unit tests
- [ ] added integration tests
- [ ] verified on staging environment (screenshot attached)

### CODE REVIEW

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 1]
```diff
@@ -14,14 +17,29 @@
...
 from databricks.labs.ucx.hive_metastore import TablesCrawler
 from databricks.labs.ucx.hive_metastore.mapping import Rule, TableMapping
 from databricks.labs.ucx.hive_metastore.tables import MigrationCount, Table, What
 
 logger = logging.getLogger(__name__)
 
 
+@dataclass
+class TableMigrationStatus:
+    src_workspace_id: str
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/table_migrate.py:34]: on this level, we have nothing to do with a `workspace_id`

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 2]
```diff
@@ -458,3 +480,33 @@ def _recreate_view(self, to_view_name, view_text):
         create_sql = f"CREATE VIEW {to_view_name} AS {view_text}"
         logger.info(f"Creating view {to_view_name}")
         self._backend.execute(create_sql)
+
+
+class TableStatusCrawler(CrawlerBase[TableMigrationStatus]):
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/table_migrate.py:485]:
```suggestion
class MigrationProgress(CrawlerBase[TableMigrationStatus]):
```
probably a better name

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py (hunk 1 above)
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/table_migrate.py:34]: OK

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 3]
```diff
@@ -54,6 +72,10 @@ def migrate_tables(self, *, what: What | None = None):
                 tasks.append(partial(self._migrate_table, table.src, table.rule))
         Threads.strict("migrate tables", tasks)
 
+    def get_seen_tables(self):
+        self._init_seen_tables()
```
[the remark this answers is not in the document]
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/table_migrate.py:76]: OK

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py (hunk 2 above)
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/table_migrate.py:485]: OK

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py (hunk 1 above)
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/table_migrate.py]: Removed

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/table_migrate.py:76]: Done

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py (hunk 2 above)
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/table_migrate.py]: Done

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 4]
```diff
@@ -34,6 +51,7 @@ def __init__(
         self._backend = backend
         self._ws = ws
         self._tm = table_mapping
+        self._migration_status = MigrationStatus(ws, backend, "hive_metastore", tables_crawler)
```
[the remark this answers is not in the document]
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/table_migrate.py:54]: OK

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 5]
```diff
@@ -458,3 +460,62 @@ def _recreate_view(self, to_view_name, view_text):
         create_sql = f"CREATE VIEW {to_view_name} AS {view_text}"
         logger.info(f"Creating view {to_view_name}")
         self._backend.execute(create_sql)
+
+
+class MigrationStatus(CrawlerBase[TableMigrationStatus]):
```
[the remark this answers is not in the document]
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/table_migrate.py:465]: Done

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/table_migrate.py:54]: Done

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 6]
```diff
@@ -14,14 +17,28 @@
...
+)
 from databricks.labs.ucx.hive_metastore import TablesCrawler
 from databricks.labs.ucx.hive_metastore.mapping import Rule, TableMapping
 from databricks.labs.ucx.hive_metastore.tables import MigrationCount, Table, What
 
 logger = logging.getLogger(__name__)
 
 
+@dataclass
+class TableMigrationStatus:
```
[the remark this answers is not in the document]
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/table_migrate.py:33]: Done

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 7]
```diff
@@ -458,3 +462,62 @@ def _recreate_view(self, to_view_name, view_text):
         create_sql = f"CREATE VIEW {to_view_name} AS {view_text}"
         logger.info(f"Creating view {to_view_name}")
         self._backend.execute(create_sql)
+
+
+class MigrationStatusRefresher(CrawlerBase[MigrationStatus]):
+    def __init__(self, ws: WorkspaceClient, sbe: SqlBackend, schema, table_crawler: TablesCrawler):
+        super().__init__(sbe, "hive_metastore", schema, "table_migration_status", MigrationStatus)
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/table_migrate.py:469]:
```suggestion
        super().__init__(sbe, "hive_metastore", schema, "migration_status", MigrationStatus)
```

File: src/databricks/labs/ucx/install.py [hunk 8]
```diff
@@ -162,6 +163,7 @@ def deploy_schema(sql_backend: SqlBackend, inventory_schema: str):
             functools.partial(table, "workspace_objects", WorkspaceObjectInfo),
             functools.partial(table, "permissions", Permissions),
             functools.partial(table, "submit_runs", SubmitRunInfo),
+            functools.partial(table, "table_migration_status", MigrationStatus),
```
Comment (nfx) [inline src/databricks/labs/ucx/install.py:166]:
```suggestion
            functools.partial(table, "migration_status", MigrationStatus),
```

File: tests/unit/hive_metastore/test_table_migrate.py [hunk 9]
```diff
@@ -384,7 +398,170 @@ def test_is_upgraded():
...
+def test_table_status_reset(mocker):
+    errors = {}
+    rows = {}
+    backend = MockBackend(fails_on_first=errors, rows=rows)
+    table_crawler = create_autospec(TablesCrawler)
+    client = create_autospec(WorkspaceClient)
+    table_status_crawler = MigrationStatusRefresher(client, backend, "ucx", table_crawler)
+    table_status_crawler.reset()
+    assert list(backend.queries) == [
+        "DELETE FROM hive_metastore.ucx.table_migration_status",
```
Comment (nfx) [inline tests/unit/hive_metastore/test_table_migrate.py:512]:
```suggestion
        "DELETE FROM hive_metastore.ucx.migration_status",
```

**Review by nfx — APPROVED** [review]
can we rename table before we merge it?

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py (hunk 7 above)
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/table_migrate.py]: Sure

File: src/databricks/labs/ucx/install.py (hunk 8 above)
Comment (FastLee) [reply src/databricks/labs/ucx/install.py]: Done

File: tests/unit/hive_metastore/test_table_migrate.py (hunk 9 above)
Comment (FastLee) [reply tests/unit/hive_metastore/test_table_migrate.py]: Done

**Review by nfx — APPROVED** [review]

**Review by nfx — APPROVED** [review]

**Pull request merged**